rfcs/001-mid-level-file-api.md

```diff
@@ -0,0 +1,187 @@
+
+    fn raw_mime_type(&self) -> String { ... }
+
+    fn as_raw(&self) -> &web_sys::Blob;
+}
+```
+There are two structs that implement this trait: `Blob` and `File`.
+
+```rust
+#[derive(Debug, Clone)]
+struct Blob { ... }
+
+impl Blob {
+  fn new<T>(content: T) -> Blob
+    where
+        T: std::convert::Into<BlobContents> // We'll look at BlobContents below
+    { ... }
+
+  fn new_with_options<T>(content: T, mime_type: Option<String>) -> Blob
+    where
+        T: std::convert::Into<BlobContents>
+    { ... }
+
+  fn slice(&self, start: usize, end: usize) -> Blob { ... }
+}
+
+impl From<web_sys::Blob> for Blob { ... }
+
+
+impl BlobLike for Blob { ... }
+
+#[derive(Debug, Clone)]
+pub struct File { ... }
+
+impl File {
+  fn new<T>(
+        name: String,
+        contents: Option<T>,
+        mime_type: Option<String>,
+        last_modified_date: Option<u64>,
+    ) -> File
+    where
+        T: std::convert::Into<BlobContents>,
+    { ... }
+
+  fn name(&self) -> String { ... }
+
+  fn last_modified_date(&self) -> u64 { ... }
+
+  fn slice(&self, start: usize, end: usize) -> File { ... }
+
+  fn as_blob(self) -> Blob { ... }
+}
+
+impl BlobLike for File { ... }
+```
+
+Both Blob and File come with builders that allow for creating new Blobs and Files in the builder style and not having to use `new` directly.
+
+`BlobContents` is simply a new-type around `wasm_bindgen::JsValue`s that can be used as the content of `Blob`s and `File`s:
+
+```rust
+#[derive(Debug, Clone)]
+pub struct BlobContents {
+    inner: wasm_bindgen::JsValue,
+}
+```
+
+There are there conversions from types into `BlobContents` only for the types that make sense:
+
+```rust
+impl std::convert::Into<BlobContents> for &str
+impl std::convert::Into<BlobContents> for &[u8]
+impl std::convert::Into<BlobContents> for Blob
+impl std::convert::Into<BlobContents> for js_sys::ArrayBuffer
+```
+
+Lastly there's the `FileReader` which allows reading from BlobLike objects. We'll have two implementations of this, one based on callbacks and the other based on futures.
+
+The callbacks implementation has three categories of callbacks: start, progress and read. Start and progress callbacks are directly related to the `onloadstart` and `onprogress` callbacks on `web_sys::FileReader`. The read variety of callbacks, are a combination of `onload`, `onloadend`, `onloaderror`, and `onabort`. The callback receives a result which is an error if the underlying read was aborted or errored.
+
+The futures implementation likewise exposes success, error, and abort through the fact that futures are `Result`-like. Progress events are exposed as a stream. In the future, we may expose the entire lifecycle of a read through a stream.
+
+```rust
+mod callbacks {
+  #[derive(Debug)]
+  pub struct FileReader { ... }
+
+  impl FileReader {
+    fn new() -> FileReader { ... }
+
+    fn read_as_string<F>(self, blob: &impl BlobLike, callback: F)
```

Comment:
Why do these take `&impl BlobLike` and not `A where A: BlobLike`?

Reply:
`impl Trait` is generalyl the preferred short hand unless you need to reference the type parameter in another bound. They should be equivalent, either way.

Reply:
Ah, right, I had confused it with `&dyn BlobLike`.

I still think it should use generics though, so it can be used with turbofish.